godot-ffi/src/global.rs

```diff
@@ -137,32 +121,64 @@ impl<T> Global<T> {
             }
         };
 
-        Some(GlobalGuard { mutex_guard })
+        // SAFETY: `mutex_guard` was either set to `Initialized` above, or we returned from the function.
+        Some(unsafe { GlobalGuard::new_unchecked(mutex_guard) })
     }
 }
 
 // ----------------------------------------------------------------------------------------------------------------------------------------------
 // Guards
 
-/// Guard that temporarily gives access to a `Global<T>`'s inner value.
-pub struct GlobalGuard<'a, T> {
-    mutex_guard: MutexGuard<'a, InitState<T>>,
-}
+// Encapsulate private fields.
+mod global_guard {
+    use std::ops::{Deref, DerefMut};
+    use std::sync::MutexGuard;
 
-impl<T> Deref for GlobalGuard<'_, T> {
-    type Target = T;
+    use super::InitState;
 
-    fn deref(&self) -> &Self::Target {
-        self.mutex_guard.unwrap_ref()
+    /// Guard that temporarily gives access to a `Global<T>`'s inner value.
+    pub struct GlobalGuard<'a, T> {
+        // Safety invariant: Is `Initialized`.
+        mutex_guard: MutexGuard<'a, InitState<T>>,
     }
-}
 
-impl<T> DerefMut for GlobalGuard<'_, T> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        self.mutex_guard.unwrap_mut()
+    impl<'a, T> GlobalGuard<'a, T> {
+        pub(super) fn new(mutex_guard: MutexGuard<'a, InitState<T>>) -> Option<Self> {
+            match &*mutex_guard {
+                InitState::Initialized(_) => Some(Self { mutex_guard }),
+                _ => None,
+            }
+        }
+
+        /// # Safety
+        ///
+        /// The value must be `Initialized`.
+        pub(super) unsafe fn new_unchecked(mutex_guard: MutexGuard<'a, InitState<T>>) -> Self {
+            debug_assert!(matches!(*mutex_guard, InitState::Initialized(_)));
```

Comment:
this isnt necessary as `unwrap_unchecked` panics in debug already.

Reply:
the error message is different, and this also more explicitly lays out what the reconditioned is